src/metadata.rs

```diff
@@ -54,17 +56,18 @@ pub struct Package {
   pub license: Option<String>,
   pub license_file: Option<String>,
   pub description: Option<String>,
-  pub source: Option<String>,
+  pub source: Option<SourceId>,
```

Comment:
Its probably not a good idea to change this (the definition of Metadata or its components).

Many of the metadata objects are written to match the output of the `cargo metadata` command, so you can't add or change fields in a way that `cargo metadata` itself couldn't logically support.

I don't have a better idea at the moment though.

Reply:
I'll take another pass at it. The reason it was changed was because Cargo is unable to go SourceId -> URL -> SourceId, so I was trying to preserve the original.

Reply:
Might be possible to deserialize it directly back into a SourceId --
https://github.com/rust-lang/cargo/blob/805fbebc830174ad34feb774e019769d6fe7bbe9/src/cargo/core/source/source_id.rs#L356

(which would mean you could replace `source: String` with `source: SourceId`, though this would be a _small_ avenue for a breakage if Cargo changes how the structure of SourceId).

Reply:
Turns out it was a holdover from previous attempts. What I was ultimately after was figuring out some way of identifying where a project was checked out to, and the SourceId wasn't particularly helpful.

Reply:
I lied, I did need it. Using serde_json to pass it back and forth, because I need to preserve the `precise()` to get the Git reference.